standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MsckPartitionExpressionProxy.java

```diff
@@ -44,6 +53,46 @@ public String convertExprToFilter(final byte[] exprBytes, final String defaultPa
   @Override
   public boolean filterPartitionsByExpr(List<FieldSchema> partColumns, byte[] expr, String
     defaultPartitionName, List<String> partitionNames) throws MetaException {
+    String partExpr = new String(expr, StandardCharsets.UTF_8);
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Partition expr: {}", expr);
+    }
+    //This is to find in partitionNames all that match expr
+    //reverse of the Msck.makePartExpr
+    Set<String> partValueSet = new HashSet<>();
+    String[] parts = partExpr.split(" AND ");
+    for ( String part : parts){
+      String[] colAndValue = part.split("=");
+      String key = FileUtils.unescapePathName(colAndValue[0]);
+      //take the value inside without the single quote marks '2018-10-30' becomes 2018-10-31
+      String value = FileUtils.unescapePathName(colAndValue[1].substring(1, colAndValue[1].length()-1));
+      partValueSet.add(key+"="+value);
+    }
+
+    List<String> partNamesSeq =  new ArrayList<>();
+    for (String partition : partitionNames){
+      boolean isMatch = true;
+      for ( String col : partValueSet){
+        //list of partitions [year=2001/month=1, year=2002/month=2, year=2001/month=3]
+        //Given expr: e.g. year='2001' AND month='1'. Only when all the expressions in the expr can be found,
+        //do we add the partition to the filtered result [year=2001/month=1]
+        if (partition.indexOf(col) == -1){
+          isMatch = false;
+          break;
+        }
+      }
+      if (isMatch){
+        partNamesSeq.add(partition);
+      }
+    }
+    partitionNames.clear();
+    partitionNames.addAll(partNamesSeq);
+    LOG.info("The returned partition list is of size: {}", partitionNames.size());
+    for(String s : partitionNames){
+      if (LOG.isDebugEnabled()) {
```

Comment:
This if condition can be moved outside of the iteration

Reply:
will do